cpp/inc/bond/core/parser.h

```diff
@@ -468,37 +455,19 @@ class DynamicParser
 
             UnknownField(id, type, transform);
         }
+    }
 
-        if (!_base)
-        {
-            // If we are not parsing a base class, and we still didn't get to
-            // the end of the struct, it means that:
-            //
-            // 1) Actual data in the payload had deeper hierarchy than payload schema.
-            //
-            // or
-            //
-            // 2) We parsed only part of the hierarchy because that was what
-            //    the transform "expected".
-            //
-            // In both cases we emit remaining fields as unknown
```

Comment:
There's a bug here that we probably should fix before refactoring. If we're parsing and we detect that we're in an unexpected hierarchy depth, that should cause serialization--and perhaps other transforms--to fail.

Reply:
Why do you think that needs to be fixed before this change? Also won't the fix likely be in `detail::ParserInheritance`?

Reply:
Take a look at issue https://github.com/Microsoft/bond/issues/742 for the details of the problem.